Add a Global Value Numbering Pass

Zentrik opened the pull request:
This enables common subexpression elimination, fixes #50887. This is based on the NewGVN pass in LLVM.

A possible avenue for increasing the number of eliminations would be to implement code hoisting (eg. an expression that is also present in all predecessors to a block could be moved to the immediate dominator and then replaced; alternatively we could replace it with a phi node) or Partial Redundancy Elimination.

I'm not sure when it would be optimal for this pass to run, the effects of a function may allow a non inlined call to be eliminated completely whilst when inlined only parts may be deleted.

For small functions most of the time is spent on allocating two arrays (`ssa_to_ssa` and the array backing `val_to_ssa`). For larger functions the `get!` dominates, #51091 speeds up `get!` 2x. I don't think there any easy optimizations left, one possible avenue would be computing use lists (a list of every instruction in which a `SSAValue` is used) and using that to not have to iterate over every instruction.

Zentrik committed:
Global Value Numbering Pass

Zentrik committed:
Don't replace :the_exception

Also rename VN

Zentrik committed:
Fix tests

Zentrik committed:
Fix for master

Zentrik committed:
Fix bug

Zentrik committed:
Fix test

Zentrik committed:
Add docstring and tests

Zentrik committed:
Change Int to Csize_t

Zentrik committed:
Fix Whitespace

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+        # end
+        # svec(key...)
+
+        key = ccall(:jl_alloc_svec, SimpleVector, (Csize_t,), length(stmt.args)+1);
```
Clever, but no. That said, since the only thing this is needed for is congruences, it would be fine to do this with an array, but you'd need to make sure to root all Expr args arrays for the lifetime of the computation, which is possible, but I don't think it's worth the tradeoff. I don't see why the Vector{Any} version above should be that much slower.

Keno replied:
Also, I would be fine with doing this optimization in the allocation optimization passes in the backend, it's just not the right place to hardcode here.

Keno replied:
That said, this is also of course just using an svec allocation to compute a hash key. It should be just as possible to just compute the hash key directly without the extra allocation.

Zentrik replied:
The performance hit is due to having to allocate both a Vector and a svec and then copying the Vector to the svec.  `gvn!` on `Base.code_ircode(exp, (Float64,); optimize_until="compact 2"` using the Vector{Any} implementation takes 36ms, here's the profile of this specific code
![image](https://github.com/JuliaLang/julia/assets/19725290/ecdfabbc-155d-41ff-9b1c-75f96b0ce453)
And using the unsafe svec implementation 27ms, 
![image](https://github.com/JuliaLang/julia/assets/19725290/f37b92a4-523e-4932-92a6-cabd26a29db5)

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+    while changed
+        changed = false
+
+        # Reverse Post Order traversal of dominator tree
```
Comment should mention why this is an RPO traversal (I.e. it's an IR invariant).

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+
+        # Reverse Post Order traversal of dominator tree
+        for (blockidx, block) in enumerate(ir.cfg.blocks), i in block.stmts
+            if !(ir.stmts.stmt[i] isa Expr) & !(ir.stmts.stmt[i] isa PhiNode)
```
`ir[SSAValue(i)][:stmt]` is the preferred API for this access. Also, pull it out into a variable.

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+
+            stmt::Union{Expr, PhiNode} = ir.stmts.stmt[i]
+
+            # IR_FLAG_NOTHROW is necessary, can't exclude it otherwise inference tests start to fail for some reason
```
If it's not IR_FLAG_NOTHROW, it may not be moved without affecting the semantics.

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+            end
+
+            value = if stmt isa Expr
+                if stmt.head === :the_exception # Do not eliminate exceptions
```
This should not be IR_FLAG_CONSISTENT. If it is, something earlier has a bug that should be fixed rather than special casing here.

Zentrik replied:
That's what I figured but wasn't sure. I'll try and come up with some MWE's then.

Zentrik replied:
This doesn't work for example 
```julia
function f()
    try
        throw(42)
    catch err
        err
    end
end
```
```julia
julia> ir = Base.code_ircode(f, ()) |> only |> first
1 1 ─ %1 = $(Expr(:enter, #3))                                                               │
  2 ─      Main.throw(42)::Union{}                                                           │
  └──      unreachable                                                                       │
  3 ┄      $(Expr(:leave, 1))                                                                │
  4 ─ %5 = $(Expr(:the_exception))::Any                                                      │
  │        $(Expr(:pop_exception, :(%1)))::Any                                               │
  └──      return %5                                                                         │
                                                                                              

julia> ir[SSAValue(5)][:flag]
0x00000170
```
The consistent flag is not set on `f` however, `Base.infer_effects(f) = (!c,+e,!n,+t,+s,+m,+u)`. I didn't encounter an example where the consistent flag wasn't set on a `:the_exception`.

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+        end
+        # empty!(val_to_ssa; preserve_size=true)
+        empty!(val_to_ssa)
+        sizehint!(val_to_ssa, length(ir.stmts.stmt))
```
```suggestion
        sizehint!(val_to_ssa, length(ir.stmts))
```

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+    length(stmt.values) == 0 && return nothing
+
+    no_of_edges = length(stmt.edges)
+    key = Vector{Any}(undef, no_of_edges*2 + 2)
```
I think for this to work properly, the edges need to be sorted. I don't think we currently have that invariant in the IR in general, though we probably could.

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+    allthesame = true
+    deletions = 0
+    for j in eachindex(stmt.values)
+        !isassigned(stmt.values, j) && return nothing
```
GVN refinement of missing predecessors is allowable to any value, though that can be a future enhancement.

Zentrik replied:
I tried refining them but that started causing segfaults in the sroa pass so I'll leave this for a future enhancement.

Keno commented on base/compiler/ssair/domtree.jl:
```diff
@@ -714,3 +714,44 @@ function naive_idoms(blocks::Vector{BasicBlock}, is_post_dominator::Bool=false)
+"""
+    dominates(dfsnumbers::AbstractVector{DFSNumber}, bb1::Int, bb2::Int) -> Bool
+"""
+function dominates(dfsnumbers::AbstractVector{DFSNumber}, bb1::BBNumber, bb2::BBNumber)
```
This should be some sort of `DFSCachedDomtree` type rather than a simple AbstractVector.

Keno reviewed (dismissed):
This is nice work, but doesn't work as is, because it doesn't consider the state of the heap.
For example, consider the following function:

```
function foo()
    arr = Int[0]
    a = arr[]
    arr[] = 1
    b = arr[]
    (a, b)
end
```

As is, `a`, and `b` will be considered congruent and replaced with each other. However, this is incorrect.

Keno commented:
That said, if we promote inaccessiblememonly to an IR flag, we could probably do this at least for things that don't touch memory, which might still be useful. Will have to wait until after #50805 though, since the flag space is currently full.

vtjnash commented on base/compiler/ssair/passes.jl:
```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+        # svec(key...)
+
+        key = ccall(:jl_alloc_svec, SimpleVector, (Csize_t,), length(stmt.args)+1);
+        ptr = convert(Ptr{Ptr{Cvoid}}, pointer_from_objref(key)) + sizeof(Ptr{Cvoid})
+        for (i, arg) in enumerate(stmt.args)
+            if isa(arg, SSAValue)
+                unsafe_store!(ptr, ccall(:jl_box_ssavalue, Ptr{Any}, (Csize_t,), ssa_to_ssa[arg.id]), i)
+            else
+                unsafe_store!(ptr, arrayptr(Ptr{Nothing}, stmt.args, i), i)
+            end
+        end
+        unsafe_store!(ptr, pointer_from_objref(stmt.head), length(stmt.args)+1)
```
As Keno mentioned, all of this is UB in Julia. This needs to be a mutable Vector{Any} and not an svec, and unsafe calls are forbidden.

Zentrik committed:
Address reviews

Zentrik commented:
> This is nice work, but doesn't work as is, because it doesn't consider the state of the heap. For example, consider the following function:
> 
> ```
> function foo()
>     arr = Int[0]
>     a = arr[]
>     arr[] = 1
>     b = arr[]
>     (a, b)
> end
> ```
> 
> As is, `a`, and `b` will be considered congruent and replaced with each other. However, this is incorrect.

I'm a bit confused, this seems to work fine for me. I see the same IR before the GVN pass, after the GVN pass and after "compact 2" on master (#95af5a08926).

```
julia> foo()
(0, 1)

julia> ir = Base.code_ircode(foo, (); optimize_until="GVN") |> only |> first
2 1 ── %1  = $(Expr(:foreigncall, :(:jl_alloc_array_1d), Vector{Int64}, svec(Any, Int64), 0, :(:ccall), Vector{Int64}, 1, 1))::Vector{Int64}
  │          Base.arrayset(false, %1, 0, 1)::Vector{Int64}unds_setindex!
3 │    %3  = $(Expr(:boundscheck, true))::Bool     getindex
  └───       goto #16 if not %3         ││┃         _getindex
  2 ── %5  = Base.arraysize(%1, 1)::Int64││╻╷╷       checkbounds
  └───       goto #3                    ││││╻╷╷╷      checkbounds
  3 ──       goto #4                    │││││┃││││     checkbounds_indices
  4 ──       goto #5                    ││││││┃││       all
  5 ── %9  = (%5 === 1)::Bool           │││││││╻╷        _all
  └───       goto #7 if not %9          │││││││╻         _all
  6 ──       goto #8                    ││││││││  
  7 ──       goto #9                    ││││││││  
  8 ──       goto #9                    ││││││││  
  9 ┄─ %14 = φ (#7 => false, #8 => true)::Bool│   
  └───       goto #10                   │││││││   
  10 ─       goto #11                   ││││││    
  11 ─       goto #12                   │││││     
  12 ─       goto #14 if not %14        ││││      
  13 ─       goto #15                   ││││      
  14 ─       invoke Base.throw_boundserror(%1::Vector{Int64}, ()::Tuple{})::Union{}
  └───       unreachable                ││││      
  15 ─       nothing::Nothing           │         
  16 ┄ %23 = $(Expr(:boundscheck, false))::Bool      getindex
  │    %24 = Base.arrayref(%23, %1, 1)::Int64     
  └───       goto #17                   ││╻         _getindex
  17 ─       goto #18                   ││        
4 18 ─ %27 = $(Expr(:boundscheck, true))::Bool     setindex!
  └───       goto #33 if not %27        ││┃         _setindex!
  19 ─ %29 = Base.arraysize(%1, 1)::Int64││╻╷╷       checkbounds
  └───       goto #20                   ││││╻╷╷╷      checkbounds
  20 ─       goto #21                   │││││┃││││     checkbounds_indices
  21 ─       goto #22                   ││││││┃││       all
  22 ─ %33 = (%29 === 1)::Bool          │││││││╻╷        _all
  └───       goto #24 if not %33        │││││││╻         _all
  23 ─       goto #25                   ││││││││  
  24 ─       goto #26                   ││││││││  
  25 ─       goto #26                   ││││││││  
  26 ┄ %38 = φ (#24 => false, #25 => true)::Bool  
  └───       goto #27                   │││││││   
  27 ─       goto #28                   ││││││    
  28 ─       goto #29                   │││││     
  29 ─       goto #31 if not %38        ││││      
  30 ─       goto #32                   ││││      
  31 ─       invoke Base.throw_boundserror(%1::Vector{Int64}, ()::Tuple{})::Union{}
  └───       unreachable                ││││      
  32 ─       nothing::Nothing           │         
  33 ┄ %47 = $(Expr(:boundscheck, false))::Bool      setindex!
  │          Base.arrayset(%47, %1, 1, 1)::Vector{Int64}
  └───       goto #34                   ││╻         _setindex!
  34 ─       goto #35                   ││        
5 35 ─ %51 = $(Expr(:boundscheck, true))::Bool     getindex
  └───       goto #50 if not %51        ││┃         _getindex
  36 ─ %53 = Base.arraysize(%1, 1)::Int64││╻╷╷       checkbounds
  └───       goto #37                   ││││╻╷╷╷      checkbounds
  37 ─       goto #38                   │││││┃││││     checkbounds_indices
  38 ─       goto #39                   ││││││┃││       all
  39 ─ %57 = (%53 === 1)::Bool          │││││││╻╷        _all
  └───       goto #41 if not %57        │││││││╻         _all
  40 ─       goto #42                   ││││││││  
  41 ─       goto #43                   ││││││││  
  42 ─       goto #43                   ││││││││  
  43 ┄ %62 = φ (#41 => false, #42 => true)::Bool  
  └───       goto #44                   │││││││   
  44 ─       goto #45                   ││││││    
  45 ─       goto #46                   │││││     
  46 ─       goto #48 if not %62        ││││      
  47 ─       goto #49                   ││││      
  48 ─       invoke Base.throw_boundserror(%1::Vector{Int64}, ()::Tuple{})::Union{}
  └───       unreachable                ││││      
  49 ─       nothing::Nothing           │         
  50 ┄ %71 = $(Expr(:boundscheck, false))::Bool      getindex
  │    %72 = Base.arrayref(%71, %1, 1)::Int64     
  └───       goto #51                   ││╻         _getindex
  51 ─       goto #52                   ││        
6 52 ─ %75 = Core.tuple(%24, %72)::Tuple{Int64, Int64}
  └───       return %75
```

Looking at the flags of `arrayref`,`arrayset` and `arraysize` we get `0x00000110`, `0x00000100` and `0x00000130` respectively so these don't get touched by this pass anyways (the consistent flag isn't set on any of them).

Keno commented:
Hmm, I may have been wrong, and IR_FLAG_CONSISTENT might be strong enough. I'll need to think about it again to make sure.

Zentrik committed:
Actually do something if phinode is equivalent to a SSAValue

Zentrik committed:
Revert previous commit

It is also legal for Phi node's to have missing incoming edges, in which case Phi node cannot be used if that arriving from a missing edge so this transformation is not always legal.

Keno commented on base/compiler/ssair/passes.jl:
```diff
@@ -1913,8 +1913,6 @@ function perform_symbolic_evaluation(stmt::PhiNode, ssa_to_ssa, blockidx)
     copyto!(key, no_of_edges+3, stmt.values)
 
-    firstval = nothing
-    allthesame = true # If all values into phi node are the same SSAValue
```
A missing edge is `undef` in the LLVM sense, so any value is legal. The issue is that the value may not be available on that predecessor, so you need to check dominance.

Zentrik committed:
Add back phi simplification if identical to a SSAValue

Zentrik committed:
Stylistic changes

Zentrik committed:
Add test

Zentrik committed:
Add DFSCachedDomtree type

Zentrik committed:
Fix whitespace

Zentrik committed:
Fix test

Zentrik committed:
Remove unnecessary check

Zentrik committed:
Sort PhiNodes

Zentrik committed:
Rename old reference to function

Zentrik committed:
Don't simplify PhiNode when we can't

Also fixes bug when calling dominates, where firstedge is still set to 0 and thus out of bounds.

Zentrik committed:
Fix typo

nsajko commented:
Does this fix #50887?

Zentrik commented:
> Does this fix #50887?

Yes that was the inspiration for this.

nsajko commented:
FTR, if you put a `Fixes #50887` line into a commit or PR description, Github is supposed to automatically [close](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) the issue.

Zentrik committed:
Fix Typo

Zentrik committed:
Refine missing predecessors

Also hopefully fix some indexing bugs.

Zentrik committed:
Fix typos

Zentrik committed:
Fix typo, indexing bug and add comment

Zentrik committed:
Revert Refine missing predecessors

It broke building Julia for some reason,
```
Segmentation fault
in expression starting at compiler/compiler.jl:181
perform_lifting! at ./compiler/ssair/passes.jl:667
```

Zentrik committed:
Minor optimization

Add fast path back in.

Zentrik committed:
Remove fix for consistent flag not being set on :the_exception

Zentrik committed:
Simplify code

Zentrik committed:
Change argument name

Prevent accidental use and make intent clearer

Zentrik committed:
Add comment about IR invariant

vchuravy commented on base/compiler/optimize.jl:
```diff
@@ -797,6 +797,7 @@ function run_passes_ipo_safe(
     @pass "Inlining"  ir = ssa_inlining_pass!(ir, sv.inlining, ci.propagate_inbounds)
     # @timeit "verify 2" verify_ir(ir)
     @pass "compact 2" ir = compact!(ir)
+    @pass "GVN"       ir = gvn!(ir) # needs to be placed immediately after compact!
```
When writing a comment like this, having an explanation for why is helpful.

Zentrik committed:
Use IncrementalCompact

I think this should all be legal. Unfortunately iterating over `compact` makes this pass significantly slower, ~2x, as lots of time is spent in `renumber_ssa2!`.

Zentrik committed:
Revert "Use IncrementalCompact"

This reverts commit 057d8174604b133b765d758032487a7381a912e0.
This broke building somehow and was quite slow anyways.

Zentrik committed:
Improve comments

Address review.
Fix bug in not wrapping result form `ssa_to_ssa` in a `SSAValue` before writing into `key` in `perform_symbolic_evaluation`

Zentrik committed:
Fix typo

[no ci]

Zentrik committed:
Add more tests, fix perform_symbolic_evaluation, clean up code

Fixes simplification of PhiNode to SSAValue by reverting indexing fix and setting `firstval` to equivalent SSAValue.

Zentrik committed:
Fix typo in test

Zentrik committed:
Fix tests

Zentrik committed:
Fix typo in comment

Zentrik committed:
Simplify symbolic evaluation of PhiNodes

Zentrik committed:
Hoist allocations out

Makes pass about 5% faster on `x->@inline(exp(x))/(1+@inline(exp(x))` and halves allocations

nsajko commented:
> I'm not sure when it would be optimal for this pass to run, the effects of a function may allow a non inlined call to be eliminated completely whilst when inlined only parts may be deleted.

You already give a reason why it's useful before inlining, here's a reason for having the pass run after inlining:

Suppose I'm calling a function whose implementation I can't modify, such as `sqrt`, but I want some error checks to be eliminated. In the case of `sqrt(x)` this would be a check that throws in case `x < 0`, so assume we know `0 ≤ x`. This assumption may be encoded in Julia like so: `(x < 0) && Core.Intrinsics.llvmcall("unreachable", Union{}, Tuple{})`. In this way, one is already able to eliminate the check, but only after LLVM gets involved:

```julia
function unsafe_sqrt(x)
  @inline begin
    is_negative = x < zero(x)
    is_negative &&
      Core.Intrinsics.llvmcall("unreachable", Union{}, Tuple{})
    sqrt(x)
  end
end
```

```julia-repl
julia> Base.infer_effects(unsafe_sqrt, Tuple{Float64})   # Julia thinks the function may throw
(+c,!e,!n,!t,!s,!m,!u)′

julia> @code_typed unsafe_sqrt(0.3)
CodeInfo(
1 ─ %1 = Base.lt_float(x, 0.0)::Bool
└──      goto #3 if not %1
2 ─      (Core.Intrinsics.llvmcall)("unreachable", Union{}, Tuple{})::Union{}
└──      unreachable
3 ─ %5 = Base.lt_float(x, 0.0)::Bool
└──      goto #5 if not %5
4 ─      invoke Base.Math.throw_complex_domainerror(:sqrt::Symbol, x::Float64)::Union{}
└──      unreachable
5 ─ %9 = Base.Math.sqrt_llvm(x)::Float64
└──      goto #6
6 ─      return %9
) => Float64

julia> @code_native debuginfo=:none unsafe_sqrt(0.3)
        .text
        .file   "unsafe_sqrt"
        .globl  julia_unsafe_sqrt_7199          # -- Begin function julia_unsafe_sqrt_7199
        .p2align        4, 0x90
        .type   julia_unsafe_sqrt_7199,@function
julia_unsafe_sqrt_7199:                 # @julia_unsafe_sqrt_7199
; Function Signature: unsafe_sqrt(Float64)
# %bb.0:                                # %top
        #DEBUG_VALUE: unsafe_sqrt:x <- $xmm0
        push    rbp
        mov     rbp, rsp
        vsqrtsd xmm0, xmm0, xmm0
        pop     rbp
        ret
```

So LLVM successfully optimizes the machine code to a single `vsqrtsd` operation, with no throwing, but Julia doesn't know that `unsafe_sqrt` doesn't throw. I'm thinking if there was a common-subexpression elimination pass after inlining, it could deduplicate the two `Base.lt_float(x, 0.0)::Bool` calls, after which it would be easier (although still not implemented, I guess) for Julia to eliminate the throw as unreachable.

Zentrik committed:
Fix typo

Zentrik committed:
Switch to Memory where possible

Zentrik committed:
Workaround fill! not being defined whilst bootstrapping

Zentrik committed:
Fix bootstrap issue and fixed type

Indices are Int's I believe.

Zentrik committed:
Revert switch to Memory

`sort!` is not defined for memory whilst bootstrapping.

Zentrik committed:
Fix misoptimization with phi nodes

Zentrik commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

Zentrik commented:
I think it might be a good idea to run PkgEval on this to root out any more bugs. #51091 and #53844 should speed up this pr by 2-3x.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51120#issuecomment-2227349148) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4b4d09a_vs_df3fe22/report.md).
